activerecord/lib/active_record/attribute_methods/write.rb

```diff
@@ -39,6 +39,12 @@ def write_attribute(attr_name, value)
       # This method exists to avoid the expensive primary_key check internally, without
       # breaking compatibility with the write_attribute API
       def _write_attribute(attr_name, value) # :nodoc:
+        column = self.class.column_for_attribute(attr_name)
+
+        if value.nil? && !column.null && column.has_default?
+          value = column.default
+        end
```

Comment:
Given the `This method exists to avoid the expensive primary_key check internally` comment above, this feels like the wrong place to add this logic. 

Having said that, the `#write_attribute` method wasn't invoked in the way I thought it'd be, and the call to `@attributes.write_from_user(attr_name, value)` doesn't have access to the `.column_for_attribute` method for the model class.